RFC 146 - Implement "Production Deploy Access" (#147)
alphagov/govuk-rfcs: rfc-146-production-deploy-access.md

There were [attempts at formalising this process around 2016](https://docs.google.com/document/d/1lo1JAkFIeWrIl-7bzkzps8rYi2hscV6IOpDzc8dhFLs/edit#). Unfortunately we didn't go into detail on why 2ndline shifts were required. The theory is that the logic went something like this:

- In the early days of GOV.UK our deployment pipeline and monitoring was immature
- This meant that debugging issues with deployments often required ssh access
- In turn, this meant that only people with ssh access were allowed to do deployments
- Because it's easy to make mistakes when using highly privileged access through ssh, we required people to do a couple of 2ndline shifts to make sure they knew what they were doing

Note that by this logic, the 2ndline requirements are defending against accidental mistakes, not against malicious access. We have separate defences against malicious employees (security clearance, probation etc.).

## Proposal

We should implement a new level of access - "Production Deploy Access". This level of access should include:

- Permission to deploy apps in Jenkins (ideally without Jenkins admin permission)
- Permission to merge continuously deployed applications

> - Readonly access to AWS, Fastly, logging systems, etc.

Reviewer: I fully 👍🏻 this RFC, but particularly for these points.
I like the idea of getting more read only access to things like AWS early on.

I would have liked to have been able to follow along more when pairing with others, earlier. Or had a poke around to familliarise myself, before I was able to make changes.

- In the early days of GOV.UK our deployment pipeline and monitoring was immature
- This meant that debugging issues with deployments often required ssh access
- In turn, this meant that only people with ssh access were allowed to do deployments
- Because it's easy to make mistakes when using highly privileged access through ssh, we required people to do a couple of 2ndline shifts to make sure they knew what they were doing

Note that by this logic, the 2ndline requirements are defending against accidental mistakes, not against malicious access. We have separate defences against malicious employees (security clearance, probation etc.).

## Proposal

We should implement a new level of access - "Production Deploy Access". This level of access should include:

- Permission to deploy apps in Jenkins (ideally without Jenkins admin permission)
- Permission to merge continuously deployed applications
- Readonly access to AWS, Fastly, logging systems, etc.

> - admin access to GOV.UK Signon (create and edit normal users)

Reviewer 2: Is this mainly to grant yourself permission to apps to debug etc? If so, could that be a different permission too?

This level of access should not include:

> - ssh permission to production or staging

Reviewer 3: Should there also be a restriction on deploying Puppet?

Otherwise a deploy-only user could give themselves higher levels of access (e.g. SSH into prod machines) or make changes to Jenkins tasks by deploying a branch of puppet.

Reviewer 4: >a deploy-only user could give themselves higher levels of access

This kind of depends on whether we're thinking of this as a guard against *malicious* users or simply careless/inexperienced ones.

And of course even with production deploy access they'd need someone else to review; it's not an absolute protection but should make it harder to make questionable changes.

Author: Yeah, this isn't intended to guard against malicious users - doing 2ndline shifts doesn't make you any more trustworthy.

There are plenty of ways a malicious user could use "deploy" access to do escalate to higher levels (I won't go in to them here because it's a public forum, but if it's worthwhile we could discuss them in a google doc addendum to this RFC).

The point about "they'd need someone else to review" is interesting - at the moment, the Deploy App job takes a commit-ish, so there's no protection (that I'm aware of) against someone asking Jenkins to deploy a branch (which doesn't have to be reviewed).

Reviewer 5: We have on occasion wanted to deploy a branch to staging. I know this for sure but can't recall why. Production I'm with you, though.

Author: I'm going to resolve this thread without putting any special restrictions on puppet.

I do think there's a wider piece about tightening up our access controls, but it's very tied up with the replatforming project and the move to Argo. I'm sure we can improve on what we've got now.

- admin access to the Deploy Jenkins in production or staging
- admin or poweruser access to AWS
- admin access to other systems (e.g. Fastly etc.)
- superadmin access to GOV.UK Signon (create and edit all user types and edit applications)

> Engineers should be granted this level of access once they've got the required level of security clearance and have been working on GOV.UK for at least one full month. Access should be granted at the discretion of the engineer's tech lead. Before approving access, tech leads should ensure that the engineer:

Reviewer 5: Could/should this be revoked if the tech lead makes the call too early?Should it be a consideration? If so do we need to define the rules for it? Is there anything like that for current existing prod access?

Reviewer 6: I echo Bill's point about the 1 month being arbitrary. Suggest we make this a guideline instead, e.g.:

```suggestion
Access should be granted at the discretion of the engineer's tech lead, once the engineer has the required level of security clearance. Before approving access, tech leads should ensure that the engineer:
```

Author: I don't think we've got a process for revoking access if it's granted too early - I think it should be rare enough that we can handle it exception by exception.

- is aware of our processes and standards around code review
- understands the responsibilities that releasing code brings with it
- knows how to roll back to an older release if there are any issues
- knows how to get help from someone with more access if they need it

> These rules should be the same for contractors and civil servants at all levels. Junior and apprentice developers should be given production deploy access once their tech lead feels it's appropriate.

Author: I've already removed the requirement for 1 month's tenure in 1dcc5972, so I'll make this:

```suggestion
These rules should be the same for contractors and civil servants at all levels, including juniors and apprentices.
```

The rules for gaining Production Admin Access should remain the same - engineers will still be required to complete two 2ndline shifts.

## Consequences

We will update the [Rules for Getting Production Access](https://docs.publishing.service.gov.uk/manual/rules-for-getting-production-access.html) to match the outcome of this RFC.

> We will create a new GitHub Team for the middle level of access, so we'll have "GOV.UK", "GOV.UK Deploy", and "GOV.UK Production".

Reviewer 6: I like the team names, but wonder if these should be consistent with the terminology used earlier, i.e.

```suggestion
We will create a new GitHub Team for the middle level of access, so we'll have "GOV.UK", "GOV.UK Production Deploy", and "GOV.UK Production Admin".
```

(Or use "GOV.UK Deploy" and "GOV.UK Production" throughout this RFC).

Author: I sort of agree, but I'm not sure what I might break by renaming the GitHub teams. At least [the Jenkins Access controls](https://github.com/alphagov/govuk-puppet/blob/e5f47f2d4330ff0713145f93747364921ad3e4b9/hieradata_aws/class/production/jenkins.yaml#L28) would have to be updated to include the new team names.

Can you think of anything else? I can shoot for the clearer names, and if it becomes too difficult I guess I can rename them back.

We will set up the staging and production deploy Jenkins instances to allow people in the GOV.UK Deploy GitHub Team to trigger builds.

The bottleneck of needing people with production access to merge and deploy code will be substantially widened, since it will be much easier to give people deploy access.

> Teams may be less willing to have new engineers doing 2ndline shifts, since the carrot of production access will be a bit less appetising if people can get deploy access without the shifts. We'll need to keep an eye on this, since our current position is that 2ndline shifts are independently valuable things for engineers to do.  

Reviewer 5: If this theory were correct we'd be seeing a drop off of engineer availability after they'd done the minimum for access. Are we? I think that typically teams are reluctant to have engineers do 2nd line shifts because it's disruptive but reluctantly support the model because there's no real alternative.

Author: Yeah, I don't think we do see that, so perhaps this is an unhelpful hypothetical. I'd be happy to remove this paragraph (but let's leave it for a few days incase anyone else has thoughts on this).

Reviewer 6: Agree this paragraph should be removed. Our 2nd line slots aren't volunteer-only, they're mandatory and allocated, so how a team feels about doing shifts shouldn't be relevant.
